Sources/MongoSwift/BSON/Document.swift

```diff
@@ -227,6 +227,18 @@ extension Document {
             self.storage = DocumentStorage(fromPointer: self.data)
         }
     }
+
+    /// If the document already has an _id, returns it as-is. Otherwise, returns a new document
+    /// containing all the keys from this document, with an _id prepended.
+    internal func withID() throws -> Document {
+        if self.hasKey("_id") {
+            return self
+        }
+
+        var idDoc: Document = ["_id": ObjectId()]
```

Comment:
Would it be worth it performance-wise to add functionality (internally) for prepending to documents without constructing an entirely new one?

Reply:
so, we might want something like that is when `CollectionType != Document`. 
in those cases, we create a new document when we call `BSONEncoder.encode()`, and then we immediately create a second document here to prepend the `_id`. (when `CollectionType == Document`, the encoder no-ops and returns the same document you passed in. then we have to make a new document anyway in order to avoid modifying user input.) 

however, I think it might actually make more sense to implement that with an internal variant of `encode` that allows you to provide an `inout` starter doc into which we encode the struct. that would also make our [encoding for index creation](https://github.com/mongodb/mongo-swift-driver/blob/fd95efb2d7e6360d706ec227c2c54977eaa049d9/Sources/MongoSwift/MongoCollection%2BIndexes.swift#L221) a bit cleaner - libmongoc expects everything to be in a single struct so we have to make two encode calls and then merge the resulting docs. 

did you have any other ideas in mind for where this would be useful?

Reply:
Ah, right. So no gain in the `Document` case. 

And the index encoding is the only other place I can find as well where this type of thing would be useful. I don't think we really encode stuff piecemeal and combine it in any other places. Anyways, for this ticket I think were fine as-is can can re-evaluate later (filed [SWIFT-336](https://jira.mongodb.org/browse/SWIFT-336)).

Reply:
sounds good,thanks for filing that!